src/promptflow/promptflow/batch/_batch_engine.py

```diff
@@ -158,67 +151,112 @@ def run(
+                apply_default_value_for_input(self._flow.inputs, each_line_input) for each_line_input in batch_inputs
+            ]
+            run_id = run_id or str(uuid.uuid4())
 
-    async def _exec_batch_internal(
+            # execute lines
+            if isinstance(self._executor_proxy, PythonExecutorProxy):
+                line_results.extend(self._executor_proxy._exec_batch(batch_inputs, output_dir, run_id))
+            else:
+                await self._exec_batch(line_results, batch_inputs, run_id)
+            handle_line_failures([r.run_info for r in line_results], raise_on_line_failure)
+
+            # execute aggregation nodes
+            aggr_exec_result = await self._exec_aggregation(batch_inputs, line_results, run_id)
+            # use the execution result to update aggr_result to make sure we can get the aggr_result in _exec_in_task
+            self._update_aggr_result(aggr_result, aggr_exec_result)
+
+            # persist outputs to output dir
+            outputs = [
+                {LINE_NUMBER_KEY: r.run_info.index, **r.output}
+                for r in line_results
+                if r.run_info.status == Status.Completed
+            ]
+            self._persist_outputs(outputs, output_dir)
+            # summary some infos from line results and aggr results to batch result
+            return BatchResult.create(self._start_time, datetime.utcnow(), line_results, aggr_result)
+        except asyncio.CancelledError:
+            # If the batch run is canceled, log it and ignore the exception.
+            bulk_logger.warning("The batch run is canceled.")
+
+    async def _exec_batch(
         self,
+        line_results: List[LineResult],
         batch_inputs: List[Mapping[str, Any]],
         run_id: Optional[str] = None,
     ) -> List[LineResult]:
-        line_results = []
         total_lines = len(batch_inputs)
-        # TODO: concurrent calls to exec_line instead of for loop
-        for i, each_line_input in enumerate(batch_inputs):
-            # TODO: catch line run failed to avoid one line break others
-            line_result = await self._executor_proxy.exec_line_async(each_line_input, i, run_id=run_id)
-            for node_run in line_result.node_run_infos.values():
-                self._storage.persist_node_run(node_run)
-            self._storage.persist_flow_run(line_result.run_info)
-            line_results.append(line_result)
-            # log the progress of the batch run
-            log_progress(self._start_time, bulk_logger, len(line_results), total_lines)
-        return line_results
+        completed_line = 0
+        pending = [
+            asyncio.create_task(self._executor_proxy.exec_line_async(line_inputs, i, run_id))
+            for i, line_inputs in enumerate(batch_inputs)
+        ]
 
-    async def _exec_aggregation_internal(
+        while completed_line < total_lines:
+            try:
+                async with asyncio.Semaphore(DEFAULT_CONCURRENCY):
```

Comment:
I remember all these tasks would be pending execution when we start task, will this Semaphore works here? Do we have tests for this?

Reply:
I put Semaphore outside of _exec_batch. I'll try to test if it works.

Reply:
Updated and tested it works.